### DISCUSSION TITLE
Login Page Static Rendering + Browser Save Credentials

### DISCUSSION THREAD

**Discussion by Irrelon** [vercel/next.js#37169] [Help]
Hey folks,

I've got a login page that currently uses `getServerSideProps` so that the login form on the page can submit to itself (POST /login) which enables the browser to offer the user to save the username and password credentials.

Having `getServerSideProps` obviously slows down page loading compared to a pure static page.

Is there a way to achieve a login with a POST to an API route e.g. /api/login but still get the browser to save login credentials?

The /api/login route would probably need to do a redirect back to the application route in order to serve the app to the user after they have logged in but in my travels I have been led to understand that a 302 redirect will cause the browser to not show a "save credentials" flow.

Any ideas?

One such flow I came up with was to POST to another page such as /login/process or something like that, and then that page can use client-side routing to redirect the user, and since the user doesn't hit that page initially to login, the actual /login page could still be statically rendered.

Am I on the right track with that idea?

**Comment by icyJoseph** [comment]
That's a very nice trick! Btw, you could also take a look at https://next-runtime.meijer.ws/getting-started/1-introduction, which abstract this away nicely.

With that being said, if you made `api/login`, and need redirects, you better not use `fetch` because the browser only uses cookies applies the cookies, when it is the `document` that initiated redirect.

However, you can do:

```jsx
const LoginForm = () => {
    return (
        <form
            method="POST"
            action="/api/login/"
        >
            {/* other necessary fields */}
            <button type="submit">
                Login
            </button>
        </form>
    );
};
```

And then from the `api/login` endpoint you can tell the browser to move elsewhere.

I reckon, you said "save credentials" specifically, do you mean, as in: https://web.dev/security-credential-management-save-forms/
